File: gpstaffapp/src/components/StudentSelector.js
import React, { useState, useMemo } from 'react';
import {
  View,
  Text,
  TextInput,
  TouchableOpacity,
  ScrollView,
  ActivityIndicator,
} from 'react-native';
import { Ionicons } from '@expo/vector-icons';

export default function StudentSelector({
  selectedStudentId,
  onSelectStudent,
  studentOptions = {},
  loading = false,
  onSearch,
  searchValue = '',
  setSearchValue,
  sectionOptions = [],
  selectedSection = '',
  onSelectSection,
}) {
  const [isExpanded, setIsExpanded] = useState(!selectedStudentId);
  const [localSearch, setLocalSearch] = useState('');

  // Convert studentOptions { id: name } or array to entries
  const studentList = useMemo(() => {
    if (!studentOptions) return [];
    if (Array.isArray(studentOptions)) {
      return studentOptions.map((s) => ({
        id: s?.student_id || s?.id || '',
        name: s?.student_name || s?.name || '',
      }));
    }
    return Object.entries(studentOptions).map(([id, name]) => ({
      id: id || '',
      name: name || '',
    }));
  }, [studentOptions]);

  // Filter students locally by search text
  const filteredStudents = useMemo(() => {
    const query = (localSearch || searchValue || '').toLowerCase().trim();
    if (!query) return studentList;
    return studentList.filter((s) => {
      const idStr = s?.id != null ? String(s.id).toLowerCase() : '';
      const nameStr = s?.name != null ? String(s.name).toLowerCase() : '';
      return idStr.includes(query) || nameStr.includes(query);
    });
  }, [studentList, localSearch, searchValue]);

  const selectedStudentName = useMemo(() => {
    if (!selectedStudentId) return '';
    if (studentOptions && studentOptions[selectedStudentId]) {
      return studentOptions[selectedStudentId];
    }
    const found = studentList.find(
      (s) => s?.id != null && String(s.id) === String(selectedStudentId)
    );
    return found ? found.name : String(selectedStudentId);
  }, [selectedStudentId, studentOptions, studentList]);

  return (
    <View className="mb-4">
      <Text className="text-xs font-bold text-gray-600 mb-2 uppercase">
        Target Student *
      </Text>

      {/* Selected Student Display (When collapsed and student selected) */}
      {selectedStudentId && !isExpanded ? (
        <View className="bg-primary/5 border border-primary/20 rounded-2xl p-3.5 flex-row items-center justify-between shadow-sm">
          <View className="flex-row items-center flex-1 pr-2">
            <View className="w-10 h-10 rounded-xl bg-primary/10 items-center justify-center mr-3">
              <Ionicons name="person" size={20} color="#2b66a2" />
            </View>
            <View className="flex-1">
              <Text className="text-xs font-bold text-primary">
                ID: {String(selectedStudentId)}
              </Text>
              <Text className="text-sm font-semibold text-gray-800" numberOfLines={1}>
                {selectedStudentName || 'Student Selected'}
              </Text>
            </View>
          </View>
          <View className="flex-row items-center">
            <TouchableOpacity
              onPress={() => setIsExpanded(true)}
              className="bg-primary/10 px-3 py-1.5 rounded-lg mr-1.5 active:bg-primary/20"
            >
              <Text className="text-xs font-bold text-primary">Change</Text>
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => {
                onSelectStudent('');
                setIsExpanded(true);
              }}
              className="bg-red-50 p-1.5 rounded-lg active:bg-red-100"
            >
              <Ionicons name="close" size={16} color="#ef4444" />
            </TouchableOpacity>
          </View>
        </View>
      ) : (
        <View className="bg-gray-50 border border-gray-200 rounded-2xl p-4 shadow-sm">
          {/* Header row in expanded box */}
          <View className="flex-row items-center justify-between mb-3">
            <View className="flex-row items-center">
              <Ionicons name="search" size={16} color="#2b66a2" style={{ marginRight: 6 }} />
              <Text className="text-xs font-bold text-gray-700">Search & Select Student</Text>
            </View>
            {selectedStudentId ? (
              <TouchableOpacity
                onPress={() => setIsExpanded(false)}
                className="bg-gray-200/80 px-2.5 py-1 rounded-lg"
              >
                <Text className="text-[11px] font-bold text-gray-600">Done</Text>
              </TouchableOpacity>
            ) : null}
          </View>

          {/* Search Input Bar */}
          <View className="bg-white border border-gray-200 rounded-xl px-3 py-2 flex-row items-center shadow-sm mb-3">
            <Ionicons name="search" size={16} color="#9ca3af" />
            <TextInput
              value={localSearch}
              onChangeText={(txt) => {
                setLocalSearch(txt);
                if (setSearchValue) setSearchValue(txt);
                if (onSearch) onSearch(txt);
              }}
              placeholder="Type name or ID to filter..."
              placeholderTextColor="#9ca3af"
              className="flex-1 text-xs text-gray-800 ml-2 py-0"
              autoCapitalize="none"
              autoCorrect={false}
            />
            {localSearch.length > 0 && (
              <TouchableOpacity
                onPress={() => {
                  setLocalSearch('');
                  if (setSearchValue) setSearchValue('');
                  if (onSearch) onSearch('');
                }}
              >
                <Ionicons name="close-circle" size={16} color="#9ca3af" />
              </TouchableOpacity>
            )}
          </View>

          {/* Section Filter Chips */}
          {sectionOptions && sectionOptions.length > 0 && (
            <View className="mb-3">
              <ScrollView horizontal showsHorizontalScrollIndicator={false} className="flex-row">
                <TouchableOpacity
                  onPress={() => {
                    if (onSelectSection) onSelectSection('');
                  }}
                  className={`mr-1.5 px-3 py-1 rounded-full border ${
                    !selectedSection
                      ? 'bg-primary border-primary'
                      : 'bg-white border-gray-200'
                  }`}
                >
                  <Text
                    className={`text-[11px] font-bold ${
                      !selectedSection ? 'text-white' : 'text-gray-600'
                    }`}
                  >
                    All Sections
                  </Text>
                </TouchableOpacity>
                {sectionOptions.map((sec) => {
                  const active = selectedSection === sec;
                  return (
                    <TouchableOpacity
                      key={sec}
                      onPress={() => {
                        if (onSelectSection) onSelectSection(sec);
                      }}
                      className={`mr-1.5 px-3 py-1 rounded-full border ${
                        active
                          ? 'bg-primary border-primary'
                          : 'bg-white border-gray-200'
                      }`}
                    >
                      <Text
                        className={`text-[11px] font-bold ${
                          active ? 'text-white' : 'text-gray-600'
                        }`}
                      >
                        {sec}
                      </Text>
                    </TouchableOpacity>
                  );
                })}
              </ScrollView>
            </View>
          )}

          {/* Matching Students List */}
          {loading ? (
            <View className="py-6 items-center justify-center">
              <ActivityIndicator size="small" color="#ff7700" />
              <Text className="text-[11px] text-gray-400 mt-2">Loading students...</Text>
            </View>
          ) : filteredStudents.length > 0 ? (
            <ScrollView
              nestedScrollEnabled={true}
              style={{ maxHeight: 220 }}
              showsVerticalScrollIndicator={true}
              className="border border-gray-100 rounded-xl bg-white p-1"
            >
              {filteredStudents.map((item) => {
                const isSelected =
                  selectedStudentId != null &&
                  item?.id != null &&
                  String(selectedStudentId) === String(item.id);
                return (
                  <TouchableOpacity
                    key={item.id}
                    onPress={() => {
                      onSelectStudent(item.id);
                      setIsExpanded(false);
                    }}
                    activeOpacity={0.7}
                    className={`p-2.5 mb-1 rounded-xl flex-row items-center justify-between ${
                      isSelected
                        ? 'bg-primary/10 border border-primary/30'
                        : 'bg-white border border-gray-50'
                    }`}
                  >
                    <View className="flex-row items-center flex-1 pr-2">
                      <View
                        className={`w-7 h-7 rounded-lg items-center justify-center mr-2 ${
                          isSelected ? 'bg-primary' : 'bg-gray-100'
                        }`}
                      >
                        <Ionicons
                          name="person"
                          size={14}
                          color={isSelected ? '#fff' : '#64748b'}
                        />
                      </View>
                      <View className="flex-1">
                        <Text
                          className={`text-xs font-bold ${
                            isSelected ? 'text-primary' : 'text-gray-800'
                          }`}
                          numberOfLines={1}
                        >
                          {item.name}
                        </Text>
                        <Text className="text-[10px] text-gray-400">
                          ID: {item.id}
                        </Text>
                      </View>
                    </View>
                    {isSelected ? (
                      <Ionicons name="checkmark-circle" size={16} color="#2b66a2" />
                    ) : (
                      <Ionicons name="chevron-forward" size={14} color="#d1d5db" />
                    )}
                  </TouchableOpacity>
                );
              })}
            </ScrollView>
          ) : (
            <View className="py-6 items-center justify-center">
              <Text className="text-xs text-gray-500 font-medium">No matching students</Text>
              <Text className="text-[10px] text-gray-400 mt-0.5">Try searching with a different term</Text>
            </View>
          )}
        </View>
      )}
    </View>
  );
}

File: gpstaffapp/lib/widgets/student_selector.dart
import 'dart:async';
import 'package:flutter/material.dart';
import '../theme/app_theme.dart';
import '../models/student_model.dart';

class StudentSelector extends StatefulWidget {
  final String selectedStudentId;
  final Function(String studentId) onSelectStudent;
  final Map<String, String> studentOptions;
  final bool loading;
  final Function(String query)? onSearch;
  final String searchValue;
  final Function(String value)? onSearchChanged;
  final List<String> sectionOptions;
  final String selectedSection;
  final Function(String section)? onSelectSection;

  const StudentSelector({
    super.key,
    required this.selectedStudentId,
    required this.onSelectStudent,
    this.studentOptions = const {},
    this.loading = false,
    this.onSearch,
    this.searchValue = '',
    this.onSearchChanged,
    this.sectionOptions = const [],
    this.selectedSection = '',
    this.onSelectSection,
  });

  @override
  State<StudentSelector> createState() => _StudentSelectorState();
}

class _StudentSelectorState extends State<StudentSelector> {
  late bool _isExpanded;
  late TextEditingController _searchController;
  Timer? _debounceTimer;

  @override
  void initState() {
    super.initState();
    _isExpanded = widget.selectedStudentId.isEmpty;
    _searchController = TextEditingController(text: widget.searchValue);
  }

  @override
  void didUpdateWidget(covariant StudentSelector oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.selectedStudentId != widget.selectedStudentId && widget.selectedStudentId.isNotEmpty) {
      _isExpanded = false;
    }
    if (oldWidget.searchValue != widget.searchValue && widget.searchValue != _searchController.text) {
      _searchController.text = widget.searchValue;
    }
  }

  @override
  void dispose() {
    _debounceTimer?.cancel();
    _searchController.dispose();
    super.dispose();
  }

  void _onSearchInput(String text) {
    setState(() {});
    widget.onSearchChanged?.call(text);

    _debounceTimer?.cancel();
    _debounceTimer = Timer(const Duration(milliseconds: 350), () {
      widget.onSearch?.call(text);
    });
  }

  List<StudentItem> get _filteredStudents {
    final query = _searchController.text.trim().toLowerCase();
    final all = widget.studentOptions.entries
        .map((e) => StudentItem(id: e.key, name: e.value))
        .toList();

    if (query.isEmpty) return all;
    return all.where((s) {
      final idMatch = s.id.toLowerCase().contains(query);
      final nameMatch = s.name.toLowerCase().contains(query);
      return idMatch || nameMatch;
    }).toList();
  }

  String get _selectedStudentName {
    if (widget.selectedStudentId.isEmpty) return '';
    return widget.studentOptions[widget.selectedStudentId] ?? widget.selectedStudentId;
  }

  @override
  Widget build(BuildContext context) {
    final students = _filteredStudents;

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const Text(
          'TARGET STUDENT *',
          style: TextStyle(
            fontSize: 11,
            fontWeight: FontWeight.bold,
            color: AppColors.textSecondary,
            letterSpacing: 0.5,
          ),
        ),
        const SizedBox(height: 8),

        if (widget.selectedStudentId.isNotEmpty && !_isExpanded)
          // Collapsed Selected Card
          Container(
            padding: const EdgeInsets.all(12),
            decoration: BoxDecoration(
              color: AppColors.primary.withOpacity(0.06),
              borderRadius: BorderRadius.circular(16),
              border: Border.all(color: AppColors.primary.withOpacity(0.25)),
            ),
            child: Row(
              children: [
                Container(
                  width: 42,
                  height: 42,
                  decoration: BoxDecoration(
                    color: AppColors.primary.withOpacity(0.12),
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: const Icon(Icons.person, color: AppColors.primary, size: 22),
                ),
                const SizedBox(width: 12),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        'ID: ${widget.selectedStudentId}',
                        style: const TextStyle(
                          fontSize: 11,
                          fontWeight: FontWeight.bold,
                          color: AppColors.primary,
                        ),
                      ),
                      Text(
                        _selectedStudentName,
                        style: const TextStyle(
                          fontSize: 14,
                          fontWeight: FontWeight.bold,
                          color: AppColors.textPrimary,
                        ),
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                      ),
                    ],
                  ),
                ),
                // Action Buttons
                TextButton(
                  onPressed: () => setState(() => _isExpanded = true),
                  style: TextButton.styleFrom(
                    backgroundColor: AppColors.primary.withOpacity(0.1),
                    padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
                    shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
                    minimumSize: Size.zero,
                  ),
                  child: const Text(
                    'Change',
                    style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: AppColors.primary),
                  ),
                ),
                const SizedBox(width: 6),
                IconButton(
                  onPressed: () {
                    widget.onSelectStudent('');
                    setState(() => _isExpanded = true);
                  },
                  icon: const Icon(Icons.close, color: AppColors.error, size: 18),
                  style: IconButton.styleFrom(
                    backgroundColor: AppColors.error.withOpacity(0.08),
                    padding: const EdgeInsets.all(6),
                    minimumSize: Size.zero,
                  ),
                ),
              ],
            ),
          )
        else
          // Expanded Search & Filter Card
          Container(
            padding: const EdgeInsets.all(16),
            decoration: BoxDecoration(
              color: const Color(0xFFF8FAFC),
              borderRadius: BorderRadius.circular(20),
              border: Border.all(color: AppColors.border),
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                // Header with Count Badge
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Row(
                      children: [
                        const Icon(Icons.search, size: 16, color: AppColors.primary),
                        const SizedBox(width: 6),
                        const Text(
                          'Search & Filter Students',
                          style: TextStyle(
                            fontSize: 12,
                            fontWeight: FontWeight.bold,
                            color: AppColors.textPrimary,
                          ),
                        ),
                        if (!widget.loading && students.isNotEmpty) ...[
                          const SizedBox(width: 8),
                          Container(
                            padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                            decoration: BoxDecoration(
                              color: AppColors.primary.withOpacity(0.1),
                              borderRadius: BorderRadius.circular(10),
                            ),
                            child: Text(
                              '${students.length}',
                              style: const TextStyle(
                                fontSize: 10,
                                fontWeight: FontWeight.bold,
                                color: AppColors.primary,
                              ),
                            ),
                          ),
                        ],
                      ],
                    ),
                    if (widget.selectedStudentId.isNotEmpty)
                      GestureDetector(
                        onTap: () => setState(() => _isExpanded = false),
                        child: Container(
                          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
                          decoration: BoxDecoration(
                            color: Colors.grey.shade200,
                            borderRadius: BorderRadius.circular(8),
                          ),
                          child: const Text(
                            'Done',
                            style: TextStyle(fontSize: 11, fontWeight: FontWeight.bold, color: AppColors.textSecondary),
                          ),
                        ),
                      ),
                  ],
                ),
                const SizedBox(height: 12),

                // Search Input Field
                Container(
                  decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(12),
                    border: Border.all(color: AppColors.border),
                  ),
                  child: TextField(
                    controller: _searchController,
                    onChanged: _onSearchInput,
                    textInputAction: TextInputAction.search,
                    onSubmitted: (val) {
                      _debounceTimer?.cancel();
                      widget.onSearch?.call(val);
                    },
                    style: const TextStyle(fontSize: 13, color: AppColors.textPrimary),
                    decoration: InputDecoration(
                      hintText: 'Search by student name or ID...',
                      hintStyle: const TextStyle(fontSize: 12, color: AppColors.textMuted),
                      prefixIcon: const Icon(Icons.search, size: 18, color: AppColors.primary),
                      suffixIcon: _searchController.text.isNotEmpty
                          ? IconButton(
                              icon: const Icon(Icons.cancel, size: 16, color: AppColors.textMuted),
                              onPressed: () {
                                _searchController.clear();
                                _onSearchInput('');
                              },
                            )
                          : null,
                      border: InputBorder.none,
                      enabledBorder: InputBorder.none,
                      focusedBorder: InputBorder.none,
                      contentPadding: const EdgeInsets.symmetric(horizontal: 12, vertical: 12),
                    ),
                  ),
                ),
                const SizedBox(height: 10),

                // Section Filter Chips
                if (widget.sectionOptions.isNotEmpty) ...[
                  SingleChildScrollView(
                    scrollDirection: Axis.horizontal,
                    physics: const BouncingScrollPhysics(),
                    child: Row(
                      children: [
                        Padding(
                          padding: const EdgeInsets.only(right: 6),
                          child: ChoiceChip(
                            label: const Text('All Sections'),
                            labelStyle: TextStyle(
                              fontSize: 11,
                              fontWeight: FontWeight.bold,
                              color: widget.selectedSection.isEmpty ? Colors.white : AppColors.textSecondary,
                            ),
                            selected: widget.selectedSection.isEmpty,
                            selectedColor: AppColors.primary,
                            backgroundColor: Colors.white,
                            side: BorderSide(
                              color: widget.selectedSection.isEmpty ? AppColors.primary : AppColors.border,
                            ),
                            shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
                            onSelected: (_) => widget.onSelectSection?.call(''),
                          ),
                        ),
                        ...widget.sectionOptions.map(
                          (sec) => Padding(
                            padding: const EdgeInsets.only(right: 6),
                            child: ChoiceChip(
                              label: Text(sec),
                              labelStyle: TextStyle(
                                fontSize: 11,
                                fontWeight: FontWeight.bold,
                                color: widget.selectedSection == sec ? Colors.white : AppColors.textSecondary,
                              ),
                              selected: widget.selectedSection == sec,
                              selectedColor: AppColors.primary,
                              backgroundColor: Colors.white,
                              side: BorderSide(
                                color: widget.selectedSection == sec ? AppColors.primary : AppColors.border,
                              ),
                              shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
                              onSelected: (_) => widget.onSelectSection?.call(sec),
                            ),
                          ),
                        ),
                      ],
                    ),
                  ),
                  const SizedBox(height: 10),
                ],

                // Student List
                if (widget.loading)
                  const Padding(
                    padding: EdgeInsets.symmetric(vertical: 24),
                    child: Center(
                      child: Column(
                        children: [
                          SizedBox(
                            width: 24,
                            height: 24,
                            child: CircularProgressIndicator(strokeWidth: 2, color: AppColors.fanta),
                          ),
                          SizedBox(height: 8),
                          Text(
                            'Loading students...',
                            style: TextStyle(fontSize: 11, color: AppColors.textMuted),
                          ),
                        ],
                      ),
                    ),
                  )
                else if (students.isNotEmpty)
                  Container(
                    constraints: const BoxConstraints(maxHeight: 220),
                    decoration: BoxDecoration(
                      color: Colors.white,
                      borderRadius: BorderRadius.circular(14),
                      border: Border.all(color: AppColors.borderLight),
                    ),
                    child: ListView.separated(
                      shrinkWrap: true,
                      padding: const EdgeInsets.all(6),
                      itemCount: students.length,
                      separatorBuilder: (_, __) => const SizedBox(height: 4),
                      itemBuilder: (context, index) {
                        final item = students[index];
                        final isSelected = widget.selectedStudentId == item.id;

                        return InkWell(
                          onTap: () {
                            widget.onSelectStudent(item.id);
                            setState(() => _isExpanded = false);
                          },
                          borderRadius: BorderRadius.circular(10),
                          child: Container(
                            padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
                            decoration: BoxDecoration(
                              color: isSelected ? AppColors.primary.withOpacity(0.08) : Colors.white,
                              borderRadius: BorderRadius.circular(10),
                              border: Border.all(
                                color: isSelected ? AppColors.primary.withOpacity(0.3) : AppColors.borderLight,
                              ),
                            ),
                            child: Row(
                              children: [
                                Container(
                                  width: 32,
                                  height: 32,
                                  decoration: BoxDecoration(
                                    color: isSelected ? AppColors.primary : Colors.grey.shade100,
                                    borderRadius: BorderRadius.circular(8),
                                  ),
                                  child: Icon(
                                    Icons.person,
                                    size: 16,
                                    color: isSelected ? Colors.white : AppColors.textSecondary,
                                  ),
                                ),
                                const SizedBox(width: 10),
                                Expanded(
                                  child: Column(
                                    crossAxisAlignment: CrossAxisAlignment.start,
                                    children: [
                                      Text(
                                        item.name,
                                        style: TextStyle(
                                          fontSize: 12,
                                          fontWeight: FontWeight.bold,
                                          color: isSelected ? AppColors.primary : AppColors.textPrimary,
                                        ),
                                        maxLines: 1,
                                        overflow: TextOverflow.ellipsis,
                                      ),
                                      Text(
                                        'ID: ${item.id}',
                                        style: const TextStyle(fontSize: 10, color: AppColors.textMuted),
                                      ),
                                    ],
                                  ),
                                ),
                                if (isSelected)
                                  const Icon(Icons.check_circle, size: 18, color: AppColors.primary)
                                else
                                  const Icon(Icons.chevron_right, size: 16, color: AppColors.textMuted),
                              ],
                            ),
                          ),
                        );
                      },
                    ),
                  )
                else
                  const Padding(
                    padding: EdgeInsets.symmetric(vertical: 20),
                    child: Center(
                      child: Column(
                        children: [
                          Icon(Icons.person_search_outlined, size: 28, color: AppColors.textMuted),
                          SizedBox(height: 6),
                          Text(
                            'No matching students',
                            style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: AppColors.textSecondary),
                          ),
                          SizedBox(height: 2),
                          Text(
                            'Try searching with a different name or ID',
                            style: TextStyle(fontSize: 11, color: AppColors.textMuted),
                          ),
                        ],
                      ),
                    ),
                  ),
              ],
            ),
          ),
        const SizedBox(height: 16),
      ],
    );
  }
}
